Machine Config Proposal

csrwng opened the pull request:
Instead of repeating the proposed types in a doc, this proposal includes code changes to represent the initial proposed type definitions.

The goal is to come to a basic consensus on an initial implementation and generate some discussion around the open issues. Like anything else, this is expected to evolve as we evolve the implementation.

mvladev commented on docs/proposal/machine-config.md:
```diff
@@ -0,0 +1,65 @@
+
+- No AWS SDK types are included. For the following reasons:
+  - Clients would also need to have the AWS SDK vendored.
+  - The type would be tied to a specific AWS client version.
```
This 100% correct and valid, but we'll provide only subset of configurations, which you can do, when launching a machine.

For example in our project we decided to more or less copy the structs which were used in aws-client-go. https://github.com/gardener/machine-controller-manager/blob/5976512b001db0b029a8982dc93e1065f099e4a7/pkg/apis/machine/v1alpha1/types.go#L763-L885 . I'm not 100% sure that our approach is 100% valid as well.

csrwng replied:
Can we adopt the approach of including as much of the API as we need for what we initially support and as the actuator evolves add more if we need to?

frobware replied:
I think this is really the best and most pragmatic choice.

mvladev commented on docs/proposal/machine-config.md:
```diff
@@ -0,0 +1,65 @@
+Cluster configuration for the AWS provider is out of scope for this proposal.
+
+## AWSMachineProviderConfig
+Defines the spec portion of a Machine specific to AWS. It is meant to be embedded in the
```
Are we focusing this API only for on-demand AWS instances? If we want to have spot instances support then we might want to include additional fields.

csrwng replied:
I would say that the we could initially focus on on-demand (because it's simpler), but we could eventually evolve to support spot instances.

csrwng commented on cloud/aws/providerconfig/types.go:
```diff
@@ -17,19 +17,112 @@ import (
+
+	// UserDataSecret contains a local reference to a secret that contains the
+	// UserData to apply to the instance
+	UserDataSecret *corev1.LocalObjectReference
```
a LocalObjectReference only contains a string (Name). But it is the more semantically correct type given that we are referencing another object in storage. See https://github.com/kubernetes/community/blob/master/contributors/devel/api-conventions.md#references-to-related-objects
You can find a few examples in the core types like this one:
https://github.com/kubernetes/kubernetes/blob/master/pkg/apis/core/types.go#L1256

csrwng commented on cloud/aws/providerconfig/types.go:
```diff
@@ -17,19 +17,112 @@ import (
+
+	// LoadBalancerIDs is the IDs of the load balancers to which the new instance
+	// should be added once it's created.
+	LoadBalancerIDs []string
```
Excellent question. I'm not sure that it's doable for machines that are part of the control plane, where we run into a chicken and egg problem trying to communicate with the control plane. However, I'm starting to think that maybe that's the responsibility of the cluster actuator and not the machine actuator. For everything else,  I think the LoadBalancer type of service should work. Wdyt?

csrwng commented:
Thank you @mvladev for the review and great suggestions!

csrwng commented:
@dgoodwin @abutcher @joelddiaz @twiest

dgoodwin commented on cloud/aws/providerconfig/types.go:
```diff
@@ -14,22 +14,151 @@
 }
 
+// AWSMachineProviderStatus is the type that will be embedded in a Machine.Status.ProviderStatus field.
+// It containsk AWS-specific status information.
```
-k.

enxebre commented on cloud/aws/providerconfig/types.go:
```diff
@@ -14,22 +14,151 @@
+// AWSMachineProviderConfig is the type that will be embedded in a Machine.Spec.ProviderConfig field
+// for an AWS instance.
 // +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
 type AWSMachineProviderConfig struct {
```
do we want to include an optional KeyName for ssh?

csrwng replied:
definitely, thx!

csrwng replied:
PR updated

ingvagabund commented on cloud/aws/providerconfig/types.go:
```diff
@@ -14,22 +14,154 @@
+	Subnet AWSResourceReference
+
+	// Placement indicates where to create the instance in AWS
+	Placement Placement
```
I like this.

frobware commented on cloud/aws/providerconfig/types.go:
```diff
@@ -14,22 +14,154 @@
+	// Subnet is a reference to the subnet to use for this instance
+	Subnet AWSResourceReference
+
+	// Placement indicates where to create the instance in AWS
```
It's more than just "indication", no? How about `specifies where...`?

frobware commented on cloud/aws/providerconfig/types.go:
```diff
@@ -14,22 +14,154 @@
+	Placement Placement
+
+	// LoadBalancerIDs is the IDs of the load balancers to which the new instance
+	// should be added once it's created.
```
nit: "it is created"

frobware commented on cloud/aws/providerconfig/types.go:
```diff
@@ -14,22 +14,154 @@
+	LoadBalancerIDs []string
+}
+
+// AWSResourceReference is a reference to a specific AWS resource by ID, ARN, or filters
```
How do you discriminate between the various choices? Which (of the optional) take precedence?

frobware commented on cloud/aws/providerconfig/types.go:
```diff
@@ -14,22 +14,154 @@
+
+// Placement indicates where to create the instance in AWS
+type Placement struct {
+	// Region is the region to use to create the insntance
```
typo: "create the instance".

frobware commented on cloud/aws/providerconfig/types.go:
```diff
@@ -14,22 +14,154 @@
+	// LastProbeTime is the last time we probed the condition.
+	// +optional
+	LastProbeTime metav1.Time
+	// LastTransitionTime is the last time the condition transitioned from one status to another.
```
Is this "from one _state_ to another"?

csrwng replied:
It's from one status to the other ... referring to the Status field above... when it transitions from corev1.StatusTrue to corev1.StatusFalse or viceversa.

csrwng commented on cloud/aws/providerconfig/types.go:
```diff
@@ -14,22 +14,154 @@
+	// Name of the filter
+	Name string
+
+	// Values includes one or more filter values
```
It's coming from the EC2 AWS API which says the same thing. I don't believe that specifying a filter name without a value is a valid selector. So it should at least have one value.

mvladev commented on cloud/aws/providerconfig/types.go:
```diff
@@ -14,22 +14,156 @@
+
+	// DeviceIndex is the index of the device on the instance for the network interface attachment.
+	// Defaults to 0.
+	DeviceIndex int
```
this should be `*int` (optional)

ingvagabund replied:
`ec2.InstanceNetworkInterfaceSpecification` wants `*int64`

chuckha replied:
I think it would be cleaner to remove this for the MVP and add it back in later when we need it.

errordeveloper commented on cloud/aws/providerconfig/types.go:
```diff
@@ -14,22 +14,156 @@
+	SecurityGroups []AWSResourceReference
+
+	// Subnet is a reference to the subnet to use for this instance
+	Subnet AWSResourceReference
```
For this to map to an ASG, then you would want a slice here.

ingvagabund replied:
Can a cluster have more subnets? If so, do we want to allow it? The subnet(s) can be automatically detected based on cluster's vpc ID.

errordeveloper replied:
I think there confusion around what this struct represents, as I pointed out above, hence the meaning (and type) of `Subnet` field is unclear.

errordeveloper commented on cloud/aws/providerconfig/types.go:
```diff
@@ -14,22 +14,156 @@
+// AWSMachineProviderConfig is the type that will be embedded in a Machine.Spec.ProviderConfig field
+// for an AWS instance.
 // +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
 type AWSMachineProviderConfig struct {
```
Is this supposed to represent a single machine? What about ASGs?

mvladev replied:
@errordeveloper single machine - we won't use ASG

errordeveloper replied:
I though this subproject wasn't going to be so prescriptive, and it would be up to any given flavour to make a decision.

detiber replied:
@errordeveloper That is a goal for the overall project, yes. That said, I think there is an inherent impedance mismatch between a Machine and an ASG-managed instance that we probably need to solve more generically in the great cluster-api.

To me it feels like the MachineSet controller would be a better manager for ASG-backed instances rather than the Machine controller. I am interested on hearing more of your thoughts on this topic, though.

errordeveloper commented on cloud/aws/providerconfig/types.go:
```diff
@@ -14,22 +14,156 @@
+	ARN *string
+
+	// Filters is a set of filters used to identify a resource
+	Filters []Filter
```
Is this mostly used for requests inputs?

csrwng replied:
Yes

errordeveloper replied:
Perhaps, it'd make more sense to use a separate struct for inputs and outputs. That's what AWS SDK does and it's more convenient to not create a additional model that one has to thing about.

ingvagabund commented on cloud/aws/providerconfig/types.go:
```diff
@@ -14,22 +14,156 @@
 
 // +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
 type AWSClusterProviderConfig struct {
 	metav1.TypeMeta `json:",inline"`
```
No need for the `json...` annotation part given all the data types are internal.

timothysc commented:
/assign @detiber 

Can you review please, and we can try to get @csrwng unblocked.

chuckha commented on cloud/aws/providerconfig/types.go:
```diff
@@ -14,22 +14,156 @@
+
+	// UserDataSecret contains a local reference to a secret that contains the
+	// UserData to apply to the instance
+	UserDataSecret *corev1.LocalObjectReference
```
I think this is outside the scope of the MVP

chuckha commented on cloud/aws/providerconfig/types.go:
```diff
@@ -17,19 +17,112 @@ import (
+
+	// CredentialsSecret is a reference to the secret with AWS credentials. Otherwise, defaults to permissions
+	// provided by attached IAM role where the actuator is running.
+	CredentialsSecret *corev1.LocalObjectReference
```
Each machine shouldn't need the aws credentials, just the actuator. I think we can remove this

chuckha commented on cloud/aws/providerconfig/types.go:
```diff
@@ -14,22 +14,156 @@
+	CredentialsSecret *corev1.LocalObjectReference
+
+	// KeyName is the name of the KeyPair to use for SSH
+	KeyName *string
```
Is SSH required? I'd be in favor of nixing SSH until we *really* need it and I don't think we'll be needing it for the MVP

chuckha commented on cloud/aws/providerconfig/types.go:
```diff
@@ -14,22 +14,156 @@
+
+	// PublicIP specifies whether the instance should get a public IP. If not present,
+	// it should use the default of its subnet.
+	PublicIP *bool
```
The MVP will be either entirely public or entirely private so each machine does not need this configuration

chuckha commented on cloud/aws/providerconfig/types.go:
```diff
@@ -14,22 +14,156 @@
+	Subnet AWSResourceReference
+
+	// Placement specifies where to create the instance in AWS
+	Placement Placement
```
Placement is out of scope for MVP, this can go away for now

chuckha commented on cloud/aws/providerconfig/types.go:
```diff
@@ -14,22 +14,156 @@
+
+	// LoadBalancerIDs is the IDs of the load balancers to which the new instance
+	// should be added once it is created.
+	LoadBalancerIDs []string
```
I think this will live on the cluster object which the machine actuator has access to

chuckha commented on cloud/aws/providerconfig/types.go:
```diff
@@ -14,22 +14,156 @@
+// AWSResourceReference is a reference to a specific AWS resource by ID, ARN, or filters.
+// Only one of ID, ARN or Filters may be specified. Specifying more than one will result in
+// a validation error.
+type AWSResourceReference struct {
```
This seems like it could get messy? How does this fit into the MVP?

csrwng replied:
It allows you to specify references to things that might not exist yet when you create your machine/machineset config. For example, you may know what your security groups will be named, but you don't have id's for them yet. Your config can use a Filter for the name tag for each of your security groups.

csrwng replied:
So far all of the resources that we reference from the machine config have a corresponding Describe call that allows them to be looked up by the fields specified in the struct. I would say we could introduce separate types of references in the future if/when we need them. Is that ok?

detiber replied:
Sounds good, thanks for the clarification.

chuckha commented on cloud/aws/providerconfig/types.go:
```diff
@@ -14,22 +14,156 @@
+}
+
+// Placement indicates where to create the instance in AWS
+type Placement struct {
```
out of scope for MVP

chuckha commented:
To give a little more context here, I'm more than happy to keep this PR as is if this work has resources, but right now it might be a little cumbersome for the agreed upon MVP. If you're happy with it and want to continue with the PR as is, we can definitely make it work.

csrwng commented:
@chuckha Thank you for the review. I'm ok with paring it down if it'll reduce friction for an initial implementation and as we add features/function, add things back into it.

chuckha commented:
That sounds good to me -- we should track the stuff that we nix with a ticket. I expect the backlog to get properly populated with tasks for the MVP this week and I don't want us to forget to add things back

csrwng commented:
@chuckha Addressed your comments. The Placement struct included the region for the machine. Are we assuming that in an initial implementation the actuator would have a default region that it uses?

chuckha commented:
/lgtm

csrwng commented:
squashed and updated versioned types

mvladev commented on cloud/aws/providerconfig/types.go:
```diff
@@ -14,25 +14,132 @@
+	// added by default by the actuator. These tags are additive. The actuator will ensure
+	// these tags are present, but will not remove any other tags that may exist on the
+	// instance.
+	Tags []TagSpecification
```
Isn't it better for this type to be `map[string]string`?
AWS client accepts `map[string]string` already and with this implementation, you have to loop over every single item to put them in a `map`

csrwng replied:
The only reason I did it this way is: https://github.com/kubernetes/community/blob/master/contributors/devel/api-conventions.md#lists-of-named-subobjects-preferred-over-maps
But if there are no objections to reverting it back to a map, then I can do so.

csrwng replied:
Actually now that I read the guidelines again... this is a pure map ... not a map of subobjects, so it doesn't make sense to make it an array.

detiber commented on cloud/aws/providerconfig/types.go:
```diff
@@ -14,25 +14,132 @@
+	LastTransitionTime metav1.Time
+	// Reason is a unique, one-word, CamelCase reason for the condition's last transition.
+	// +optional
+	Reason string
```
Why wouldn't we bubble these up to the MachineStatus ErrorReason/ErrorMessage instead of having a separate Reason/Message here?

csrwng replied:
Conditions are meant to provide more detailed status, giving you time, condition, etc. and allowing the possibility of reporting multiple conditions at a time. If we want to provide a more concise status at the higher level, then I can add an ErrorReason and ErrorMessage there.

detiber replied:
There is already an ErrorReason/ErrorMessage on the MachineStatus object: https://github.com/kubernetes-sigs/cluster-api/blob/051f338bdacb76117d73a86258bc58c946add7b5/pkg/apis/cluster/v1alpha1/machine_types.go#L112-L133

It sounds like there could be a use case for duplicating the Reason/Message here as well for more granularity, though.

csrwng commented on cloud/aws/providerconfig/types.go:
```diff
@@ -14,25 +14,132 @@
+
+	// DeviceIndex is the index of the device on the instance for the network interface attachment.
+	// Defaults to 0.
+	DeviceIndex int
```
Sounds good, will remove

csrwng commented on cloud/aws/providerconfig/types.go:
```diff
@@ -14,25 +14,132 @@
+	// added by default by the actuator. These tags are additive. The actuator will ensure
+	// these tags are present, but will not remove any other tags that may exist on the
+	// instance.
+	Tags []TagSpecification
```
will do

detiber commented on cloud/aws/providerconfig/types.go:
```diff
@@ -14,25 +14,132 @@
+}
+
+// Filter is a filter used to identify an AWS resource
+type Filter struct {
```
Makes sense, can we add a link to that to the comments to make it more clear for new users?

errordeveloper replied:
Why do we redefine the type here? Why not alias https://docs.aws.amazon.com/sdk-for-go/api/service/ec2/#Filter?

errordeveloper commented on cloud/aws/providerconfig/types.go:
```diff
@@ -14,25 +14,123 @@
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 )
 
+// AWSMachineProviderConfig is the type that will be embedded in a Machine.Spec.ProviderConfig field
```
This doesn't actually say anything about what this struct is supposed to represent. Is it a single running instance, or template of an instance or what? The comment must clarify that, at the very least. Instead it talks about relationship between structs, which is something one can figure by reading the code anyway.

errordeveloper replied:
Also, I think it's very reasonable to ask where this should be just an alias to [`ec2.Instance`](https://docs.aws.amazon.com/sdk-for-go/api/service/ec2/#Instance)?

csrwng commented:
@detiber @errordeveloper thank you for your comments. I've now updated the PR.

csrwng commented:
@errordeveloper I mentioned the reasons for not using the EC2 types directly in the proposal doc. Having separate types also gives us some freedom to choose the EC2 client we use. For example, at some point we may want to switch to the [V2 client](https://aws.amazon.com/blogs/developer/aws-sdk-for-go-2-0-developer-preview/).

detiber commented on cloud/aws/providerconfig/types.go:
```diff
@@ -14,25 +14,140 @@
+	// instance. These security groups would be set in addition to any security groups defined
+	// at the cluster level or in the actuator.
+	// +optional
+	SecurityGroups []AWSResourceReference
```
Can this be renamed to AdditionalSecurityGroups to align with the updated comments?

csrwng commented:
done

detiber commented:
/lgtm

@errordeveloper are you able to attend the implementation meeting (https://groups.google.com/d/msg/kubernetes-sig-cluster-lifecycle/__XIKigkxkA/j1BHBdznDwAJ https://docs.google.com/document/d/10dq54Fd-xa6P5Iy3p46VY1YTFqugGMd1PygDIpuRw6c/edit#) to discuss your questions/concerns further?

detiber commented:
/approve

errordeveloper commented:
I would like to attend regularly, but it clashes with general implementer's
office hours, which I am currently chairing. Let's discuss how meetings
align/misalign in the WG call today.

On Tue, 21 Aug 2018, 7:41 pm Jason DeTiberus, <notifications@github.com>
wrote:

> /lgtm
>
> @errordeveloper <https://github.com/errordeveloper> are you able to
> attend the implementation meeting (
> https://groups.google.com/d/msg/kubernetes-sig-cluster-lifecycle/__XIKigkxkA/j1BHBdznDwAJ
> https://docs.google.com/document/d/10dq54Fd-xa6P5Iy3p46VY1YTFqugGMd1PygDIpuRw6c/edit#)
> to discuss your questions/concerns further?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes-sigs/cluster-api-provider-aws/pull/4#issuecomment-414779770>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAPWSxE-ZcPCn413aWWrD0sC-2JcK3Yfks5uTFRhgaJpZM4Vs_4l>
> .
>

errordeveloper commented:
As far as I am aware, the V2 client is not very different with respect to
these struct are implemented.

On Tue, 21 Aug 2018, 7:17 pm Cesar Wong, <notifications@github.com> wrote:

> @errordeveloper <https://github.com/errordeveloper> I mentioned the
> reasons for not using the EC2 types directly in the proposal doc. Having
> separate types also gives us some freedom to choose the EC2 client we use.
> For example, at some point we may want to switch to the V2 client
> <https://aws.amazon.com/blogs/developer/aws-sdk-for-go-2-0-developer-preview/>
> .
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes-sigs/cluster-api-provider-aws/pull/4#issuecomment-414771955>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAPWS7eF3uPkbi_iynXpfvnxRmEtAbaDks5uTE67gaJpZM4Vs_4l>
> .
>
